### PR TITLE
Adds `submit_notebook` to user facing API

### PR DESCRIPTION
**Pull request by idantene**
Adds the ability to send the notebook to the agent without interacting with the CLI.
Notebook does not freeze but will not get output from the run either.
- [x] Add tests
- [x] Add documentation

### CODE REVIEW

File: meeshkan/api/utils.py [hunk 1]
```diff
@@ -1,3 +1,35 @@
...
+import requests
 
-__all__ = []  # type: List[str]
+from ..__utils__ import _get_api
+
+__all__ = ["run_notebook"]
+
+
+def run_notebook():
+    try:  # Verifies this was run in an IPython with a non-terminal kernel
```
Comment (ksaaskil) [inline meeshkan/api/utils.py]: Perhaps add the docstring?

File: meeshkan/api/utils.py [hunk 2]
```diff
@@ -1,3 +1,35 @@
...
+        ip = get_ipython()
+        if ip.__class__.__name__ != "ZMQInteractiveShell":  # Used for communication with the IPyKernel
+            return  # Not ran from JupyterNotebook or similar
```
Comment (ksaaskil) [inline meeshkan/api/utils.py]: Is this business as expected or should raise exception?

File: meeshkan/api/utils.py [hunk 3]
```diff
@@ -1,3 +1,35 @@
...
+            print("Skipping notebook server on port {port} as it's password-protected".format(port=srv['port']))
+            continue
+        if srv['token']:
+            url += "?token={token}".format(token=srv["token"])
+
+        sessions = requests.get(url).json()
+        for sess in sessions:
+            if sess['kernel']['id'] == kernel_id:
+                path = os.path.join(srv['notebook_dir'], sess['notebook']['path'])
+                _get_api().submit((path,))  # Submit notebook
```
Comment (ksaaskil) [inline meeshkan/api/utils.py]: Could perhaps use `Service.api()` in my opinion as most other package-level methods use that as well?

**Comment by idantene** [comment]
> Looks good (after adding some documentation), no idea what it does or how it works but looks good :)

**tl;dr** `meeshkan submit <this notebook>` from the notebook itself.

It's a follow-up for #133 and is the third missing piece for full NB integration. The idea being that inside a notebook, you can simply call `meeshkan.submit_notebook(...)`. The notebook will be compiled to `.py` automatically, given a name (via arguments passed to `submit_notebook`), and run as a standalone script. It's a non-blocking operation, fits into scheduling and hopefully enables fast prototyping.

File: meeshkan/api/utils.py [hunk 4]
```diff
@@ -1,3 +1,138 @@
...
+                                          connection_file_function=ipykernel.get_connection_file,
+                                          notebook_password=notebook_password)
+        if path is not None:
+            return Service.api().submit((path,), name=job_name, poll_interval=poll_interval)  # Submit notebook
+    except ValueError:  # Ran from ipython but not from jupyter notebook -> expected behaviour
+        print("submit_notebook(): Not run from notebook interpreter; ignoring...")
+        return None
+    # In theory, should never get here...
+    raise RuntimeError("Something went terribly wrong; Meeshkan couldn't locate the matching notebook server! Contact"
+                       " Meeshkan development (dev@meeshkan.com) if you see this message.")
```
Comment (ksaaskil) [inline meeshkan/api/utils.py]: Not related to this PR, but I think we should decide max two channels where users can report issues, I'd suggest meeshkan-community Slack channel and GitHub issues.

File: meeshkan/api/utils.py [hunk 5]
```diff
@@ -1,3 +1,138 @@
...
+    Submits the current notebook to the Meeshkan agent. Requires the agent to be running.
+    Can only be called from within a notebook instance.
+    On password-protected notebooks, the `password` argument must be supplied.
+    """
+    try:
+        try:
+            get_ipython_func = get_ipython
+        except NameError:
+            get_ipython_func = None
+        path = _get_notebook_path_generic(get_ipython_function=get_ipython_func,
```
Comment (ksaaskil) [inline meeshkan/api/utils.py]: Nice to see this wrapped in a function :)

File: meeshkan/api/utils.py [hunk 6]
```diff
@@ -1,3 +1,138 @@
...
+                                          list_servers_function=notebookapp.list_running_servers,
+                                          connection_file_function=ipykernel.get_connection_file,
+                                          notebook_password=notebook_password)
+        if path is not None:
+            return Service.api().submit((path,), name=job_name, poll_interval=poll_interval)  # Submit notebook
+    except ValueError:  # Ran from ipython but not from jupyter notebook -> expected behaviour
+        print("submit_notebook(): Not run from notebook interpreter; ignoring...")
+        return None
+    # In theory, should never get here...
+    raise RuntimeError("Something went terribly wrong; Meeshkan couldn't locate the matching notebook server! Contact"
```
Comment (ksaaskil) [inline meeshkan/api/utils.py]: I think this is a small problem of catching exceptions and returning None, it's really difficult both for the user and us to know what went wrong if the code does not raise?

File: meeshkan/api/utils.py [hunk 7]
```diff
@@ -1,3 +1,138 @@
...
+LOGGER = logging.getLogger(__file__)
+
+def submit_notebook(job_name: str = None, poll_interval: Optional[float] = None, notebook_password: str = None):
+    """
+    Submits the current notebook to the Meeshkan agent. Requires the agent to be running.
+    Can only be called from within a notebook instance.
+    On password-protected notebooks, the `password` argument must be supplied.
+    """
+    try:
+        try:
```
Comment (ksaaskil) [inline meeshkan/api/utils.py]: Not for this PR, but nested try-except is a bit hard for me at least to read...

File: meeshkan/api/utils.py [hunk 8]
```diff
@@ -1,3 +1,138 @@
...
+    :param list_servers_function: Callable that returns a list of dictionaries, describing the currently running
+                                    notebook servers
+    :param connection_file_function: Callable that returns the location to file containing the IPyKernel details
+    :param notebook_password: Password for the notebook server if needed.
+
+    :return Location to the current notebook if found, otherwise None.
+    :raises RuntimeError if get_ipython_function is None, or if invalid connection file is given
+    :raises ValueError if calling get_ipython_function returns a non-ZMQ Interactive Shell.
+    """
+    try:  # Verifies this was run in an IPython with a non-terminal kernel
```
Comment (ksaaskil) [inline meeshkan/api/utils.py]: Nitpicking, but verification could be part of another method to make this shorter.

File: meeshkan/api/utils.py [hunk 9]
```diff
@@ -1,3 +1,138 @@
...
+    # responses from Jupyter Notebook API are described here:
+    # pylint: disable=line-too-long
+    #     http://petstore.swagger.io/?url=https://raw.githubusercontent.com/jupyter/notebook/master/notebook/services/api/api.yaml
+    kernel_id = os.path.splitext(connection_file)[0].split('-', 1)[1]
+    for srv in list_servers_function():
+        sessions_url = "{url}api/sessions".format(url=srv['url'])
+        sess = _notebook_authenticated_session_or_none(base_url=srv['url'], port=srv['port'],
+                                                       notebook_password=notebook_password)
+        if sess is None:
+            continue
```
Comment (ksaaskil) [inline meeshkan/api/utils.py]: I think it's a bit hard to see here what went wrong here and why did not the code raise an exception?
